intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp

```diff
@@ -58,14 +58,23 @@ class Model {
   void init_model(const std::string& model_path, int n_predict, int n_batch, int ctx_size, int seed, int threads,
                   float repetition_penalty, int num_beams, bool do_sample, int top_k, float top_p, float temperature,
                   int min_new_tokens, float length_penalty, bool early_stopping, int n_keep, int n_discard,
-                  bool shift_roped_k, int batch_size, model_vocab::id pad_token);
+                  bool shift_roped_k, int batch_size, model_vocab::id pad_token, const std::string& memory_dtype);
   void reinit();
   std::vector<std::vector<model_token>> generate(const std::vector<std::vector<model_token>>& input_ids);
   std::vector<std::vector<model_token>> generate_tokens(const std::vector<std::vector<model_token>>& input_ids);
+  const std::vector<float>& evaluate_(const std::vector<std::vector<model_token>>& input_ids);
+  std::vector<std::vector<float>> evaluate(const std::vector<std::vector<model_token>>& input_ids) {
```

Comment:
how about returning numpy for better performance

Reply:
Great idea. 99%+ users will cast the result to numpy/torch for data procession. It also make possible to do zero-cost reshaping, while will be useful if we support multi-batch logits output in the future.

> https://stackoverflow.com/a/48483811/9817693
> > Pytorch docs:
> > 
> > If a numpy.ndarray, torch.Tensor, or torch.Storage is given, a new tensor that shares the same data is returned. If a Python sequence is given, a new tensor is created from a copy of the sequence.

Reply:
Updated
https://github.com/intel/intel-extension-for-transformers/blob/856b6cd9a05a43ed5dc7c0f40fe256cec6ad5324/intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp#L71-L79